cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,244 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/isin.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/buffer.h"
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/hashing.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+class MemoryPool;
+
+using internal::checked_cast;
+using internal::DictionaryTraits;
+using internal::HashTraits;
+
+namespace compute {
+
+#define CHECK_IMPLEMENTED(KERNEL, FUNCNAME, TYPE)                                       \
+  if (!KERNEL) {                                                                        \
+    return Status::NotImplemented(FUNCNAME, " not implemented for ", type->ToString()); \
+  }
+
+class HashBinaryKernelImpl: public BinaryKernel {
+
+    virtual Status Compute(FunctionContext* ctx, const ArrayData& left,
+                         const ArrayData& right, std::shared_ptr<ArrayData>& out) = 0;
+
+    public:
+    Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+                                                    Datum* out) override {
+
+        DCHECK_EQ(Datum::ARRAY, right.kind());
+        DCHECK_EQ(Datum::ARRAY, left.kind());
+
+        const ArrayData& left_data = *left.array();
+        const ArrayData& right_data = *right.array();
+
+        std::shared_ptr<ArrayData> result;
+        RETURN_NOT_OK(Compute(ctx, left_data , right_data, result));
+
+        out->value = std::move(result);
+        return Status::OK();
+    }
+    std::shared_ptr<DataType> out_type() const override { return boolean(); }
+};
+
+template <typename Type, typename Scalar>
+class IsInKernel : public HashBinaryKernelImpl{
```

Comment:
please provide some documentation as to strategy here.

Reply:
@fsaintjacques Can you suggest what to use for fixed array?

Reply:
The core of the algorithm (as python) is

```
for idx, elem in array:
  output[i] = (elem in memo_table)
```

What I'm saying is that often, the right operand list will be very small, in this case, you should replace the hashtable with a constant vector. You can encode this with a functor which is a template parater to the IsInKernel.

I'd say create a new jira ticket for this that depends on the benchmark ticket.

Reply:
@fsaintjacques thank you for the suggestion. I have created another JIRA ticket.